docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,245 @@
+    trace     string
+    traceHash tmbytes.HexBytes
+  )
+
+  // check if the denomination is clean or if it contains the trace info
+  if denomSplit[0] == denom {
+    baseDenom = denom
+    trace = portID + "/" + channelID +"/"
+  } else {
+    baseDenom = denomSplit[2]
+    traceHash = tmbytes.HexBytes(denomSplit[1])
+    // Get the value from the map trace hash -> denom identifiers prefix
+    trace = k.GetTrace(ctx, traceHash)
+    // prefix the identifiers to create the new trace
+    trace = portID + "/" + channelID +"/" + trace + "/"
+  }
+  
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String() + baseDenom
+  return denom
+}
+```
+
+The denomination also needs to be updated when token is received on the source chain:
+
+```golang
+func (k Keeper) UnprefixDenom(ctx Context, denom string) (denom, trace string, err error) {
+  denomSplit := strings.Split(denom, "/")
+  if denomSplit[0] == denom {
+    return denom, fmt.Errorf("denomination %s doesn't contain a prefix", denom)
+  }
+
+  baseDenom := denomSplit[2]
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> denom identifiers prefix
+  trace = k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  traceSplit := strings.Split(trace, "/")
+  if len(traceSplit) == 3 {
+    // the trace has only one portID/channelID pair
+    return baseDenom
+  }
+
+  // remove a single identifiers pair to create the new trace
+  trace = strings.Join(trace[2:], "/")
+  traceHash = tmbytes.HexBytes(tmhash.Sum(trace))
+
+  // set the value to the lookup table if not stored already
+  if !k.HasTrace(ctx, traceHash) {
+    k.SetTrace(ctx, traceHash)
+  }
+
+  denom = "ibc/"+ traceHash.String() + baseDenom
+  return denom
+}
+```
+
+```golang
+// GetTraceFromDenom 
+func (k Keeper) GetTraceFromDenom(ctx Context, denom string) (string, error) {
+  denomSplit := strings.Split(denom, "/")
+  if denomSplit[0] == denom {
+    return "", nil
+  }
+
+  traceHash := tmbytes.HexBytes(denomSplit[1])
+  // Get the value from the map trace hash -> denom identifiers prefix
+  trace := k.GetTrace(ctx, traceHash)
+  if trace == "" {
+    return "", Wrapf(ErrTraceNotFound, "denom: %s", denom)
+  }
+
+  return trace
+}
+
+```
+
+Additionally, the `SendTransfer`'s `createOutgoingPacket` call and the `OnRecvPacket` need to be updated to be retreive the trace info (using `GetTraceFromDenom`) prior to checking the correctness of the prefix.
+
+### Coin Changes
+
+The coin denomination validation will need to be updated to reflect these changes:
+
+- Clean denoms that don't have separators will maintain the original validation logic
+- Denominations with separators will need to have exactly the 3 components mentioned on the IBC denomination format above:
```

Comment:
Let's keep the denom validation simple and let the IBC module handle those details. Let's just add `/` as an acceptable character in demons. We may want to have prefixed demons for other use cases too (Regen does for sure).

Reply:
Yes, I think the `x/ibc` module should do all the IBC-related denom validation

Reply:
@aaronc should I just bump the max total denom length to 64 and accept slashes, and without the verification of the 3 components?

Reply:
updated

Reply:
> @aaronc should I just bump the max total denom length to 64 and accept slashes, and without the verification of the 3 components?

That sounds like the right approach.